### ISSUE TITLE
Proposal: Add estimatedDuration Property for ScheduledSessions

### ISSUE THREAD

**Issue by peter-dolkens** [openactive/modelling-opportunity-data#201]
## Proposer

England Athletics (RunTogether) / LTA

## Use Case

In Clubspark, we capture session durations as a range (0-30mins, 30-60mins, 60-90mins, 90+)

## Why is this not covered by existing properties?

The spec states that we must return an ISO8601 format duration, however it's also validating that this duration is non-zero, so we're unable to just provide the minimum Duration

## Please provide a link to example data

https://api.runtogether.co.uk/odi/public/courses

```
"subEvent": [
  {
    "type": "ScheduledSession",
    "url": "https://groups.runtogether.co.uk/RunningFree/Runs/Summary/ea313abf-4aad-4341-a827-da059c3ee00f",
    "startDate": "2018-10-01T17:00:00+01:00",
    "endDate": "2018-10-01T17:00:00+01:00",
    "duration": "PT0S",
    "eventStatus": "https://schema.org/EventScheduled",
    "name": "First steps"
    ...
  }
],
```

## Proposal

We'd like to propose an `estimatedDuration` field of type [QuantitativeValue](https://schema.org/QuantitativeValue) and `unitCode` of [MIN](http://wiki.goodrelations-vocabulary.org/Documentation/UN/CEFACT_Common_Codes)

This follows the pattern of other duration ranges in `schema.org` that use `QuantitativeValue`:

* https://schema.org/eligibleDuration
* https://schema.org/advanceBookingRequirement
* https://schema.org/deliveryLeadTime
* https://schema.org/durationOfWarranty

And follows the property naming pattern of `estimatedX`:

* https://schema.org/estimatedCost
* https://pending.schema.org/estimatedSalary

Validation rules should state that: if `estimatedDuration` is provided, then `duration` MUST not be provided. This ensures accurate and consistent data.

## [Beta](https://www.openactive.io/ns-beta/) property

| (Class) Property    |  Expected Type  | Description                                                         |
|---------------------|-----------------|---------------------------------------------------------------------|
| <a name="estimatedDuration"></a> ([`schema:Event`](https://schema.org/Event)) <br/>  `beta:estimatedDuration` | [`schema:QuantitativeValue`](https://schema.org/QuantitativeValue) | A property that allows an Event duration to be represented as a range (e.g. 0-30mins, 30-60mins, 60-90mins, 90+). |

## Example

```
"subEvent": [
  {
    "type": "ScheduledSession",
    "url": "https://groups.runtogether.co.uk/RunningFree/Runs/Summary/ea313abf-4aad-4341-a827-da059c3ee00f",
    "startDate": "2018-10-01T17:00:00+01:00",
    "endDate": "2018-10-01",
    "beta:estimatedDuration": {
      "type": "QuantitativeValue",
      "minValue": 0,
      "maxValue": 30,
      "unitCode": "MIN"
    },
    "eventStatus": "https://schema.org/EventScheduled",
    "name": "First steps"
    ...
  }
],
```

**Comment by dolkensp** [comment]
@nickevansuk I didn't mention `endDate` in this proposal yet, as I thought it could still be useful as an indicator of when a session MUST be completed by (e.g. Your course starts at 2pm, and SHOULD take 20-40 minutes, but it MUST be completed by 3pm)

Let me know what you think.

**Comment by nickevansuk** [comment]
Looks great! Just updated your example from `beta:exampleDuration` to `beta:estimatedDuration` (assume that was a typo?).

`endDate` being included makes sense to me, though arguably if there was an `endDate` one could also calculate the `duration` as the "maximum duration", which undermines the proposed validation rules. Good discussion to have on this.

For now I'll go ahead and add this to the beta namespace as you've specified.

**Comment by dolkensp** [comment]
> `endDate` being included makes sense to me, though arguably if there was an `endDate` one could also calculate the duration as the "maximum duration", which undermines the proposed validation rules. Good discussion to have on this.

Yeah, I think it's important to decide how we want to define `StartTime`, `EndTime`, and `estimatedDuration`. If we wish to distinguish between

* `StartTime`/`EndTime`: When a session MUST start after, and complete by
* `estimatedDuration`: How long a session MAY be

There is a difference, based on those definitions, but those may not be the definitions we wish to use for this specification.

As a real life example, if I take driving lessons, I pay by the hour, so they are ALWAYS 2 hours long, but the start time often varies between 5pm and 6pm, depending on when my instructor becomes available.

I would imagine many sports coaching sessions behave in a similar fashion.

**Comment by ldodds** [comment]
Looks good to me, I'd suggest as part of beta testing we ask others if they would find this useful, I think that discussion will help flush out the intended requirements to inform validation rules.

**Comment by nickevansuk** [comment]
In Parkrun's case they've found that including an `endTime` is problematic and actually puts people off, so they just include `startTime` on their website.

**Comment by croesus** [comment]
A potential difficulty with an optional `endDate` comes when searching for sessions that intersect a date and time. The logic for this is usually 

```startDate <= dateTimeOfInterest AND endDate >= dateTimeOfInterest```

If `endDate` is unspecified and assumed to be "anytime after `startDate`", then this search would match any event that started before the `dateTimeOfInterest`, even if that was a long time before.

That said , it looks like `endDate` is optional so the search code above would already have to deal with using `startDate + duration` as `endDate`. This proposal would add to that: 

```endDate = max(endDate, startDate+duration, startDate+estimatedDuration.maxValue)```

It might be simpler to require `endDate` as the maximum possible endpoint, but have `duration` or `estimatedDuration` take precedence when an event is displayed. Even Parkrun packs up eventually!

On a side note - @peter-dolkens example above has a date with no time as the `endDate`, which is midnight at the start of that day in ISO8601. This puts `endDate` before `startDate` and would be rejected as inconsistent data.

**Comment by peter-dolkens** [comment]
@croesus forgive my example - it appears we’ve used the naive mapping of .net DateTime to StartDate/EndDate in the sample feed.

The OA spec actually contains both Start/End Date and Start/End Time, so ideally we should be using those appropriately, with both `EndDate` and `EndTime` optional.

In reference to your search user case, ClubSpark usually works based on individual sessions when searching, and work our way up to the containing course, as a coach may have added a course that runs an entire year, but if they haven’t yet populated the upcoming sessions, then there is little point in returning them in results until that time.

**Comment by nickevansuk** [comment]
Relevant points from the summary of the [W3C call on 12 Feb 2020](https://w3c.openactive.io/meetings/2020-02-12-opportunity-specification-proposals):

- Discussion of this became complex, because of the need to balance:
  - the reluctance of some publishers to provide a `duration` or `endDate` at all on their Events, for fear of demotivating people.
  - a desire for strong validation confirming that `startDate`, `endDate`, and `duration` are not self-contradictory
  - the need to specify time ranges in order to give at least an indicative idea of the time commitment involved in participating in an activity
  - supporting applications relying upon the presence of these values for scheduling, etc.
- No satisfactory solution bridging all of these was arrived at on the call, and it was agreed to continue the discussion asynchronously on the GitHub thread.

**Comment by thill-odi** [comment]
Two possibilities:

- Introduce `indicativeDuration` from the [Routes spec](https://openactive.io/routes/#describing-indicative-durations-oa-indicativeduration), and make one of `duration` or `indicativeDuration` REQUIRED.
- retain the spec as is but improve guidance for `schedulingNote`

**Comment by thill-odi** [comment]
Notes from call 7 Oct 2020: simple solution would be just to leave this as a beta property, particularly as data consumers don't appear ready to process this at the moment.
